backend/quill/src/test/scala/io.buildo.toctoc/quill/authentication/test/QuillLoginAuthenticationDomainFlowSpec.scala

```diff
@@ -0,0 +1,131 @@
+  val login3 = Login("usernameosone", "password")
+  val subject = UserSubject("test")
+  val subject2 = UserSubject("test2")
+
+  "unregistered login credentials" should "not be accepted when exchanging for token" in {
+    authFlow.exchangeForTokens(login).futureValue shouldBe 'left
+  }
+
+  "registered login credentials" should "be accepted when exchanging for token" in {
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
+    authFlow.exchangeForTokens(login).futureValue shouldBe 'right
+  }
+
+  "token obtained by login" should "be validated" in {
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
+    val token = authFlow.exchangeForTokens(login).futureValue.right.value
+    authFlow.validateToken(token).futureValue.right.value shouldBe subject
+  }
+
+  "multiple login with same values" should "not be accepted in registration" in {
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'left
+  }
+
+  "multiple login with different values" should "be accepted in registration" in {
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
+    authFlow.registerSubjectLogin(subject2, login2).futureValue shouldBe 'right
+    val token2 = authFlow.exchangeForTokens(login2).futureValue.right.value
+    authFlow.validateToken(token2).futureValue.right.value shouldBe subject2
+  }
+
+  "single token unregistration" should "unregister only the specific token" in {
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
+    authFlow.registerSubjectLogin(subject2, login2).futureValue shouldBe 'right
+    val token = authFlow.exchangeForTokens(login).futureValue.right.value
+    val token2 = authFlow.exchangeForTokens(login2).futureValue.right.value
+    authFlow.unregisterToken(token).futureValue
+    authFlow.validateToken(token).futureValue shouldBe 'left
+    authFlow.validateToken(token2).futureValue shouldBe 'right
+  }
+
+  "token unregistration" should "unregister all subject's tokens" in {
+    authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
+    authFlow.registerSubjectLogin(subject2, login2).futureValue shouldBe 'right
+    val token = authFlow.exchangeForTokens(login).futureValue.right.value
+    val token2 = authFlow.exchangeForTokens(login).futureValue.right.value
+    val token3 = authFlow.exchangeForTokens(login2).futureValue.right.value
+    authFlow.unregisterAllSubjectTokens(subject).futureValue
+    authFlow.validateToken(token).futureValue shouldBe 'left
+    authFlow.validateToken(token2).futureValue shouldBe 'left
+    authFlow.validateToken(token3).futureValue shouldBe 'right
+  }
+
+  "single subject credentials unregistration" should "take effect" in {
+    authFlow.unregisterLogin(login).futureValue
+    authFlow.exchangeForTokens(login).futureValue shouldBe 'left
```

Comment:
Perhaps this test is missing a precondition:

```
authFlow.registerSubjectLogin(subject, login).futureValue shouldBe 'right
```

Reply:
👍 makes sense, I'll add it